// typemacro-monorepo/packages/specialize/dist/index.js
// src/index.ts
import * as ts from "typescript";
import {
  defineExpressionMacro,
  globalRegistry
} from "@typemacro/core";
var specializeMacro = defineExpressionMacro({
  name: "specialize",
  module: "@typemacro/specialize",
  description: "Create a specialized version of a generic function",
  expand(ctx, callExpr, args) {
    if (args.length < 2) {
      ctx.reportError(
        callExpr,
        "specialize() requires a function and an array of typeclass instances"
      );
      return callExpr;
    }
    const fnExpr = args[0];
    const instancesArg = args[1];
    if (!ts.isArrayLiteralExpression(instancesArg)) {
      ctx.reportError(
        callExpr,
        "specialize() second argument must be an array literal of typeclass instances"
      );
      return callExpr;
    }
    const instances = instancesArg.elements;
    const fnType = ctx.getTypeOf(fnExpr);
    const callSignatures = fnType.getCallSignatures();
    if (callSignatures.length === 0) {
      ctx.reportError(callExpr, "specialize() first argument must be callable");
      return callExpr;
    }
    const signature = callSignatures[0];
    const params = signature.getParameters();
    const numInstances = instances.length;
    const realParamCount = params.length - numInstances;
    if (realParamCount < 0) {
      ctx.reportError(
        callExpr,
        `specialize() provided ${numInstances} instances but function only has ${params.length} parameters`
      );
      return callExpr;
    }
    const factory = ctx.factory;
    const paramNames = [];
    const paramDecls = [];
    for (let i = 0; i < realParamCount; i++) {
      const paramSymbol = params[i];
      const paramName = factory.createIdentifier(paramSymbol.getName());
      paramNames.push(paramName);
      const paramType = ctx.typeChecker.getTypeOfSymbolAtLocation(
        paramSymbol,
        callExpr
      );
      const typeNode = ctx.typeChecker.typeToTypeNode(
        paramType,
        callExpr,
        ts.NodeBuilderFlags.None
      );
      paramDecls.push(
        factory.createParameterDeclaration(
          void 0,
          void 0,
          paramName,
          void 0,
          typeNode
        )
      );
    }
    const allArgs = [...paramNames, ...instances];
    const body = factory.createCallExpression(fnExpr, void 0, allArgs);
    return factory.createArrowFunction(
      void 0,
      void 0,
      paramDecls,
      void 0,
      factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
      body
    );
  }
});
var specializeInlineMacro = defineExpressionMacro({
  name: "specialize$",
  module: "@typemacro/specialize",
  description: "Inline a specialized function call",
  expand(ctx, callExpr, args) {
    if (args.length !== 1) {
      ctx.reportError(
        callExpr,
        "specialize$() requires exactly one function call expression"
      );
      return callExpr;
    }
    const innerCall = args[0];
    if (!ts.isCallExpression(innerCall)) {
      ctx.reportError(
        callExpr,
        "specialize$() argument must be a function call"
      );
      return callExpr;
    }
    const fnExpr = innerCall.expression;
    const fnType = ctx.getTypeOf(fnExpr);
    const callSignatures = fnType.getCallSignatures();
    if (callSignatures.length === 0) {
      return innerCall;
    }
    const signature = callSignatures[0];
    const typeParams = signature.getTypeParameters();
    if (!typeParams || typeParams.length === 0) {
      return innerCall;
    }
    const factory = ctx.factory;
    return factory.createCallExpression(
      innerCall.expression,
      innerCall.typeArguments,
      innerCall.arguments.slice()
    );
  }
});
var monoMacro = defineExpressionMacro({
  name: "mono",
  module: "@typemacro/specialize",
  description: "Monomorphize a generic function for specific type arguments",
  expand(ctx, callExpr, args) {
    if (args.length !== 1) {
      ctx.reportError(
        callExpr,
        "mono() requires exactly one function reference"
      );
      return callExpr;
    }
    const typeArgs = callExpr.typeArguments;
    if (!typeArgs || typeArgs.length === 0) {
      ctx.reportError(callExpr, "mono<T1, ...>() requires type arguments");
      return callExpr;
    }
    const fnExpr = args[0];
    const fnType = ctx.getTypeOf(fnExpr);
    const callSignatures = fnType.getCallSignatures();
    if (callSignatures.length === 0) {
      ctx.reportError(callExpr, "mono() argument must be a function");
      return callExpr;
    }
    const signature = callSignatures[0];
    const params = signature.getParameters();
    const factory = ctx.factory;
    const paramDecls = [];
    const paramRefs = [];
    for (const param of params) {
      const name = factory.createIdentifier(param.getName());
      paramRefs.push(name);
      const paramType = ctx.typeChecker.getTypeOfSymbolAtLocation(
        param,
        callExpr
      );
      const typeNode = ctx.typeChecker.typeToTypeNode(
        paramType,
        callExpr,
        ts.NodeBuilderFlags.None
      );
      paramDecls.push(
        factory.createParameterDeclaration(
          void 0,
          void 0,
          name,
          void 0,
          typeNode
        )
      );
    }
    const body = factory.createCallExpression(fnExpr, typeArgs, paramRefs);
    return factory.createArrowFunction(
      void 0,
      void 0,
      paramDecls,
      void 0,
      factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
      body
    );
  }
});
var inlineCallMacro = defineExpressionMacro({
  name: "inlineCall",
  module: "@typemacro/specialize",
  description: "Attempt to inline a function call at compile time",
  expand(ctx, callExpr, args) {
    if (args.length !== 1) {
      ctx.reportError(
        callExpr,
        "inlineCall() requires exactly one function call"
      );
      return callExpr;
    }
    const call = args[0];
    if (!ts.isCallExpression(call)) {
      ctx.reportError(
        callExpr,
        "inlineCall() argument must be a function call"
      );
      return callExpr;
    }
    const fnExpr = call.expression;
    const fnSymbol = ctx.typeChecker.getSymbolAtLocation(fnExpr);
    if (!fnSymbol) {
      return call;
    }
    const declarations = fnSymbol.getDeclarations();
    if (!declarations || declarations.length === 0) {
      return call;
    }
    const fnDecl = declarations[0];
    let body;
    let fnParams;
    if (ts.isFunctionDeclaration(fnDecl) && fnDecl.body) {
      body = fnDecl.body;
      fnParams = fnDecl.parameters;
    } else if (ts.isArrowFunction(fnDecl)) {
      body = fnDecl.body;
      fnParams = fnDecl.parameters;
    } else if (ts.isVariableDeclaration(fnDecl) && fnDecl.initializer && ts.isArrowFunction(fnDecl.initializer)) {
      body = fnDecl.initializer.body;
      fnParams = fnDecl.initializer.parameters;
    }
    if (!body || !fnParams) {
      return call;
    }
    if (!ts.isBlock(body)) {
      const factory = ctx.factory;
      const clonedParams = fnParams.map(
        (p) => factory.createParameterDeclaration(
          void 0,
          p.dotDotDotToken ? factory.createToken(ts.SyntaxKind.DotDotDotToken) : void 0,
          ts.isIdentifier(p.name) ? factory.createIdentifier(p.name.text) : p.name,
          p.questionToken ? factory.createToken(ts.SyntaxKind.QuestionToken) : void 0,
          p.type,
          p.initializer
        )
      );
      return factory.createCallExpression(
        factory.createParenthesizedExpression(
          factory.createArrowFunction(
            void 0,
            void 0,
            clonedParams,
            void 0,
            factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
            body
            // Body is already an expression, can reuse
          )
        ),
        void 0,
        call.arguments.slice()
      );
    }
    if (body.statements.length === 1) {
      const stmt = body.statements[0];
      if (ts.isReturnStatement(stmt) && stmt.expression) {
        const factory = ctx.factory;
        const clonedParams = fnParams.map(
          (p) => factory.createParameterDeclaration(
            void 0,
            p.dotDotDotToken ? factory.createToken(ts.SyntaxKind.DotDotDotToken) : void 0,
            ts.isIdentifier(p.name) ? factory.createIdentifier(p.name.text) : p.name,
            p.questionToken ? factory.createToken(ts.SyntaxKind.QuestionToken) : void 0,
            p.type,
            p.initializer
          )
        );
        return factory.createCallExpression(
          factory.createParenthesizedExpression(
            factory.createArrowFunction(
              void 0,
              void 0,
              clonedParams,
              void 0,
              factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
              stmt.expression
              // Reuse the expression
            )
          ),
          void 0,
          call.arguments.slice()
        );
      }
    }
    return call;
  }
});
function register() {
  globalRegistry.register(specializeMacro);
  globalRegistry.register(specializeInlineMacro);
  globalRegistry.register(monoMacro);
  globalRegistry.register(inlineCallMacro);
}
register();
export {
  inlineCallMacro,
  monoMacro,
  register,
  specializeInlineMacro,
  specializeMacro
};
