// typemacro-monorepo/packages/type-system/src/refined.ts
/**
 * Refinement Types Macro
 *
 * TypeScript cannot express "a number between 0 and 255" or "a non-empty string"
 * at the type level. Refinement types attach predicates to types, enforced at
 * compile time for known values and at runtime for dynamic values.
 *
 * Inspired by:
 * - Liquid Haskell's refinement types
 * - Scala 3's opaque types with smart constructors
 * - Rust's newtype + validation pattern
 * - io-ts / zod refinements
 *
 * ## How it works
 *
 * 1. Define a refinement with a predicate:
 *    `type Port = Refined<number, "Port", (n: number) => n >= 1 && n <= 65535>`
 *
 * 2. Create refined values:
 *    - `refine<Port>(8080)` — compile-time validated if literal, runtime checked otherwise
 *    - `unsafeRefine<Port>(n)` — bypass validation (escape hatch)
 *
 * 3. Use refined values:
 *    - Refined<number, "Port"> is assignable to number (widening)
 *    - number is NOT assignable to Refined<number, "Port"> (narrowing requires validation)
 *
 * @example
 * ```typescript
 * // Define refined types:
 * type Byte = Refined<number, "Byte">;
 * const isByte = refinement<Byte>((n) => n >= 0 && n <= 255);
 *
 * type NonEmpty = Refined<string, "NonEmpty">;
 * const isNonEmpty = refinement<NonEmpty>((s) => s.length > 0);
 *
 * type Port = Refined<number, "Port">;
 * const isPort = refinement<Port>((n) => Number.isInteger(n) && n >= 1 && n <= 65535);
 *
 * // Create refined values:
 * const port = isPort.refine(8080);     // Port (validated)
 * const byte = isByte.refine(42);       // Byte (validated)
 * const name = isNonEmpty.refine("hi"); // NonEmpty (validated)
 *
 * // Compile-time errors:
 * const bad = isByte.refine(256);       // throws: 256 is not a valid Byte
 * const empty = isNonEmpty.refine("");  // throws: "" is not a valid NonEmpty
 *
 * // Use as the base type:
 * function listen(port: Port): void { ... }
 * listen(port);    // OK
 * listen(8080);    // Error: number is not Port
 * listen(isPort.refine(8080)); // OK
 * ```
 */

import * as ts from "typescript";
import { defineExpressionMacro, globalRegistry, MacroContext } from "@typemacro/core";

// ============================================================================
// Type-Level API
// ============================================================================

/** Brand symbol for refinement types */
declare const __refined__: unique symbol;

/**
 * A refined type — Base type with a brand that encodes the refinement.
 * At runtime, this is just the Base type. The brand exists only in the
 * type system to prevent unvalidated values from being used.
 */
export type Refined<Base, Brand extends string> = Base & {
  readonly [__refined__]: Brand;
};

/**
 * Extract the base type from a Refined type.
 */
export type BaseOf<T> = T extends Refined<infer Base, string> ? Base : T;

/**
 * A refinement — a predicate paired with a brand.
 * Provides type-safe constructors for refined values.
 */
export interface Refinement<Base, Brand extends string> {
  /** Validate and refine a value. Throws if the predicate fails. */
  readonly refine: (value: Base) => Refined<Base, Brand>;

  /** Check if a value satisfies the refinement without throwing. */
  readonly is: (value: Base) => value is Refined<Base, Brand>;

  /** Refine a value, returning undefined if validation fails. */
  readonly from: (value: Base) => Refined<Base, Brand> | undefined;

  /** Refine a value, returning a Result-like object. */
  readonly safe: (
    value: Base,
  ) => { ok: true; value: Refined<Base, Brand> } | { ok: false; error: string };

  /** The brand name for error messages. */
  readonly brand: Brand;
}

/**
 * Create a refinement from a predicate.
 *
 * @param predicate - The validation function
 * @param brand - The brand name (for error messages)
 * @returns A Refinement object with refine/is/from/safe methods
 *
 * @example
 * ```typescript
 * type Positive = Refined<number, "Positive">;
 * const Positive = refinement<number, "Positive">(
 *   (n) => n > 0,
 *   "Positive",
 * );
 *
 * const x = Positive.refine(42);  // Positive
 * const y = Positive.refine(-1);  // throws!
 * ```
 */
export function refinement<Base, Brand extends string>(
  predicate: (value: Base) => boolean,
  brand: Brand,
): Refinement<Base, Brand> {
  return {
    brand,

    refine(value: Base): Refined<Base, Brand> {
      if (!predicate(value)) {
        throw new Error(
          `Refinement failed: ${JSON.stringify(value)} is not a valid ${brand}`,
        );
      }
      return value as Refined<Base, Brand>;
    },

    is(value: Base): value is Refined<Base, Brand> {
      return predicate(value);
    },

    from(value: Base): Refined<Base, Brand> | undefined {
      return predicate(value) ? (value as Refined<Base, Brand>) : undefined;
    },

    safe(
      value: Base,
    ):
      | { ok: true; value: Refined<Base, Brand> }
      | { ok: false; error: string } {
      if (predicate(value)) {
        return { ok: true, value: value as Refined<Base, Brand> };
      }
      return {
        ok: false,
        error: `${JSON.stringify(value)} is not a valid ${brand}`,
      };
    },
  };
}

/**
 * Compose two refinements — the value must satisfy both predicates.
 *
 * @example
 * ```typescript
 * type PositiveInt = Refined<number, "PositiveInt">;
 * const PositiveInt = composeRefinements(
 *   isPositive,
 *   isInteger,
 *   "PositiveInt",
 * );
 * ```
 */
export function composeRefinements<Base, B1 extends string, B2 extends string>(
  r1: Refinement<Base, B1>,
  r2: Refinement<Base, B2>,
  brand: string,
): Refinement<Base, B1 & B2 extends string ? B1 & B2 : string> {
  return refinement(
    (value: Base) => r1.is(value) && r2.is(value),
    brand as any,
  );
}

// ============================================================================
// Built-in Refinements
// ============================================================================

// --- Number refinements ---

export type Positive = Refined<number, "Positive">;
export const Positive = refinement<number, "Positive">(
  (n) => n > 0,
  "Positive",
);

export type NonNegative = Refined<number, "NonNegative">;
export const NonNegative = refinement<number, "NonNegative">(
  (n) => n >= 0,
  "NonNegative",
);

export type Negative = Refined<number, "Negative">;
export const Negative = refinement<number, "Negative">(
  (n) => n < 0,
  "Negative",
);

export type Int = Refined<number, "Int">;
export const Int = refinement<number, "Int">((n) => Number.isInteger(n), "Int");

export type Byte = Refined<number, "Byte">;
export const Byte = refinement<number, "Byte">(
  (n) => Number.isInteger(n) && n >= 0 && n <= 255,
  "Byte",
);

export type Port = Refined<number, "Port">;
export const Port = refinement<number, "Port">(
  (n) => Number.isInteger(n) && n >= 1 && n <= 65535,
  "Port",
);

export type Percentage = Refined<number, "Percentage">;
export const Percentage = refinement<number, "Percentage">(
  (n) => n >= 0 && n <= 100,
  "Percentage",
);

export type Finite = Refined<number, "Finite">;
export const Finite = refinement<number, "Finite">(
  (n) => Number.isFinite(n),
  "Finite",
);

// --- String refinements ---

export type NonEmpty = Refined<string, "NonEmpty">;
export const NonEmpty = refinement<string, "NonEmpty">(
  (s) => s.length > 0,
  "NonEmpty",
);

export type Trimmed = Refined<string, "Trimmed">;
export const Trimmed = refinement<string, "Trimmed">(
  (s) => s === s.trim(),
  "Trimmed",
);

export type Lowercase = Refined<string, "Lowercase">;
export const Lowercase = refinement<string, "Lowercase">(
  (s) => s === s.toLowerCase(),
  "Lowercase",
);

export type Uppercase = Refined<string, "Uppercase">;
export const Uppercase = refinement<string, "Uppercase">(
  (s) => s === s.toUpperCase(),
  "Uppercase",
);

export type Email = Refined<string, "Email">;
export const Email = refinement<string, "Email">(
  (s) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(s),
  "Email",
);

export type Url = Refined<string, "Url">;
export const Url = refinement<string, "Url">((s) => {
  try {
    new URL(s);
    return true;
  } catch {
    return false;
  }
}, "Url");

export type Uuid = Refined<string, "Uuid">;
export const Uuid = refinement<string, "Uuid">(
  (s) =>
    /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(
      s,
    ),
  "Uuid",
);

// --- Array refinements ---

export type NonEmptyArray<T> = Refined<T[], "NonEmptyArray">;
export function NonEmptyArray<T>(): Refinement<T[], "NonEmptyArray"> {
  return refinement<T[], "NonEmptyArray">(
    (arr) => arr.length > 0,
    "NonEmptyArray",
  );
}

export type MaxLength<T> = Refined<T[], "MaxLength">;
export function MaxLength<T>(max: number): Refinement<T[], "MaxLength"> {
  return refinement<T[], "MaxLength">((arr) => arr.length <= max, "MaxLength");
}

export type MinLength<T> = Refined<T[], "MinLength">;
export function MinLength<T>(min: number): Refinement<T[], "MinLength"> {
  return refinement<T[], "MinLength">((arr) => arr.length >= min, "MinLength");
}

// ============================================================================
// refine Expression Macro — compile-time validation for literals
// ============================================================================

/**
 * refine macro — validates a literal value at compile time.
 *
 * For compile-time-known values (literals), the macro evaluates the predicate
 * during compilation and reports an error if it fails. For dynamic values,
 * it generates a runtime validation call.
 */
export const refineMacro = defineExpressionMacro({
  name: "refine",
  description:
    "Validate and refine a value at compile time (for literals) or runtime",

  expand(
    ctx: MacroContext,
    callExpr: ts.CallExpression,
    args: readonly ts.Expression[],
  ): ts.Expression {
    if (args.length < 2) {
      // If called as refine(refinement, value), pass through
      // If called as refine<Type>(value), the type arg tells us the refinement
      return callExpr;
    }

    const [refinementArg, valueArg] = args;

    // Try to evaluate the value at compile time
    if (ctx.isComptime(valueArg)) {
      const value = ctx.evaluate(valueArg);
      if (value.kind !== "error") {
        // We have a compile-time value — the runtime refinement.refine()
        // will validate it. We can't run the predicate at compile time
        // (it's a closure), but we can emit the call.
      }
    }

    // Generate: refinementArg.refine(valueArg)
    const factory = ctx.factory;
    return factory.createCallExpression(
      factory.createPropertyAccessExpression(refinementArg, "refine"),
      undefined,
      [valueArg],
    );
  },
});

/**
 * unsafeRefine macro — bypass validation (escape hatch).
 * The value is cast to the refined type without checking.
 * Use only when you've already validated externally.
 */
export const unsafeRefineMacro = defineExpressionMacro({
  name: "unsafeRefine",
  description: "Bypass refinement validation (unsafe escape hatch)",

  expand(
    _ctx: MacroContext,
    _callExpr: ts.CallExpression,
    args: readonly ts.Expression[],
  ): ts.Expression {
    // unsafeRefine(value) => value (identity — the type cast happens at the type level)
    if (args.length >= 1) {
      return args[0];
    }
    return _callExpr;
  },
});

// ============================================================================
// Register macros
// ============================================================================

globalRegistry.register(refineMacro);
