// typemacro-monorepo/packages/transformer/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  MacroTransformer: () => MacroTransformer,
  default: () => macroTransformerFactory
});
module.exports = __toCommonJS(src_exports);
var ts = __toESM(require("typescript"), 1);
var import_core = require("@typemacro/core");
function macroTransformerFactory(program, config) {
  const verbose = config?.verbose ?? false;
  if (verbose) {
    console.log("[typemacro] Initializing transformer");
    console.log(
      `[typemacro] Registered macros: ${import_core.globalRegistry.getAll().map((m) => m.name).join(", ")}`
    );
  }
  return (context) => {
    return (sourceFile) => {
      if (verbose) {
        console.log(`[typemacro] Processing: ${sourceFile.fileName}`);
      }
      const ctx = (0, import_core.createMacroContext)(program, sourceFile, context);
      const transformer = new MacroTransformer(ctx, verbose);
      const result = ts.visitNode(
        sourceFile,
        transformer.visit.bind(transformer)
      );
      const macroDiagnostics = ctx.getDiagnostics();
      for (const diag of macroDiagnostics) {
        const start = diag.node ? diag.node.getStart(sourceFile) : 0;
        const length = diag.node ? diag.node.getWidth(sourceFile) : 0;
        const tsDiag = {
          file: sourceFile,
          start,
          length,
          messageText: `[typemacro] ${diag.message}`,
          category: diag.severity === "error" ? ts.DiagnosticCategory.Error : ts.DiagnosticCategory.Warning,
          code: 9e4,
          // Custom diagnostic code range for typemacro
          source: "typemacro"
        };
        const ctxWithDiag = context;
        if (ctxWithDiag.addDiagnostic) {
          ctxWithDiag.addDiagnostic(tsDiag);
        }
        if (verbose) {
          const prefix = diag.severity === "error" ? "ERROR" : "WARNING";
          const loc = diag.node ? ` at ${sourceFile.fileName}:${sourceFile.getLineAndCharacterOfPosition(start).line + 1}` : "";
          console.log(`[typemacro ${prefix}]${loc} ${diag.message}`);
        }
      }
      return result;
    };
  };
}
var MacroTransformer = class {
  constructor(ctx, verbose) {
    this.ctx = ctx;
    this.verbose = verbose;
  }
  additionalStatements = [];
  /**
   * Cache of resolved macro symbols for this source file.
   * Maps a ts.Symbol id to the MacroDefinition it resolves to (or null if not a macro).
   * Built lazily as identifiers are encountered.
   */
  symbolMacroCache = /* @__PURE__ */ new Map();
  // ---------------------------------------------------------------------------
  // Import-scoped macro resolution
  // ---------------------------------------------------------------------------
  /**
   * Resolve an identifier to a macro definition via import tracking.
   */
  resolveMacroFromSymbol(node, macroName, kind) {
    const symbol = this.ctx.typeChecker.getSymbolAtLocation(node);
    if (!symbol) {
      return this.fallbackNameLookup(macroName, kind);
    }
    const symbolId = symbol.id;
    if (symbolId !== void 0 && this.symbolMacroCache.has(symbolId)) {
      return this.symbolMacroCache.get(symbolId) ?? void 0;
    }
    const result = this.resolveSymbolToMacro(symbol, macroName, kind);
    if (symbolId !== void 0) {
      this.symbolMacroCache.set(symbolId, result ?? null);
    }
    return result;
  }
  /**
   * Core symbol resolution: follow aliases to find the original declaration,
   * then check if it comes from a known macro module.
   */
  resolveSymbolToMacro(symbol, macroName, kind) {
    let resolved = symbol;
    if (resolved.flags & ts.SymbolFlags.Alias) {
      try {
        resolved = this.ctx.typeChecker.getAliasedSymbol(resolved);
      } catch {
      }
    }
    const declarations = resolved.getDeclarations();
    if (!declarations || declarations.length === 0) {
      return this.fallbackNameLookup(macroName, kind);
    }
    for (const decl of declarations) {
      const sourceFile = decl.getSourceFile();
      const fileName = sourceFile.fileName;
      const moduleSpecifier = this.resolveModuleSpecifier(fileName);
      if (moduleSpecifier) {
        const exportName = resolved.name;
        const macro = import_core.globalRegistry.getByModuleExport(
          moduleSpecifier,
          exportName
        );
        if (macro && macro.kind === kind) {
          return macro;
        }
        if (exportName !== macroName) {
          const macroByName = import_core.globalRegistry.getByModuleExport(
            moduleSpecifier,
            macroName
          );
          if (macroByName && macroByName.kind === kind) {
            return macroByName;
          }
        }
      }
    }
    return this.fallbackNameLookup(macroName, kind);
  }
  /**
   * Map a file path back to a module specifier like "typemacro" or "@typemacro/units".
   */
  resolveModuleSpecifier(fileName) {
    const normalized = fileName.replace(/\\/g, "/");
    const nodeModulesMatch = normalized.match(
      /\/node_modules\/((?:@[^/]+\/)?[^/]+)/
    );
    if (nodeModulesMatch) {
      const pkgName = nodeModulesMatch[1];
      if (pkgName.startsWith("@typemacro/")) {
        return pkgName;
      }
      if (pkgName === "typemacro") {
        return "typemacro";
      }
      return pkgName;
    }
    if (normalized.includes("/packages/units/")) return "@typemacro/units";
    if (normalized.includes("/packages/effect-do/"))
      return "@typemacro/effect-do";
    if (normalized.includes("/packages/sql/")) return "@typemacro/sql";
    if (normalized.includes("/packages/strings/")) return "@typemacro/strings";
    if (normalized.includes("/packages/cats/")) return "@typemacro/cats";
    if (normalized.includes("/packages/comptime/")) return "@typemacro/comptime";
    if (normalized.includes("/packages/reflect/")) return "@typemacro/reflect";
    if (normalized.includes("/packages/derive/")) return "@typemacro/derive";
    if (normalized.includes("/packages/operators/"))
      return "@typemacro/operators";
    if (normalized.includes("/packages/typeclass/"))
      return "@typemacro/typeclass";
    if (normalized.includes("/packages/specialize/"))
      return "@typemacro/specialize";
    if (normalized.includes("/packages/core/")) return "@typemacro/core";
    if (normalized.includes("/packages/typemacro/")) return "typemacro";
    if (normalized.includes("/src/use-cases/units/")) return "@typemacro/units";
    if (normalized.includes("/src/use-cases/effect-do/"))
      return "@typemacro/effect-do";
    if (normalized.includes("/src/use-cases/sql/")) return "@typemacro/sql";
    if (normalized.includes("/src/use-cases/strings/"))
      return "@typemacro/strings";
    if (normalized.includes("/src/index.") || normalized.includes("/src/macros/") || normalized.includes("/src/core/") || normalized.includes("/dist/")) {
      return "typemacro";
    }
    return void 0;
  }
  /**
   * Fall back to name-based lookup for macros without module requirement.
   */
  fallbackNameLookup(name, kind) {
    let macro;
    switch (kind) {
      case "expression":
        macro = import_core.globalRegistry.getExpression(name);
        break;
      case "attribute":
        macro = import_core.globalRegistry.getAttribute(name);
        break;
      case "derive":
        macro = import_core.globalRegistry.getDerive(name);
        break;
      case "tagged-template":
        macro = import_core.globalRegistry.getTaggedTemplate(name);
        break;
      case "type":
        macro = import_core.globalRegistry.getType(name);
        break;
      case "labeled-block":
        macro = import_core.globalRegistry.getLabeledBlock(name);
        break;
    }
    if (macro?.module) {
      return void 0;
    }
    return macro;
  }
  /**
   * Visit a node and potentially transform it
   */
  visit(node) {
    if (ts.isSourceFile(node) || ts.isBlock(node) || ts.isModuleBlock(node)) {
      return this.visitStatementContainer(node);
    }
    const transformed = this.tryTransform(node);
    if (transformed !== void 0) {
      return transformed;
    }
    return ts.visitEachChild(
      node,
      this.visit.bind(this),
      this.ctx.transformContext
    );
  }
  /**
   * Visit a node that contains a statement list.
   */
  visitStatementContainer(node) {
    const statements = Array.from(node.statements);
    const newStatements = [];
    let modified = false;
    for (let i = 0; i < statements.length; i++) {
      const stmt = statements[i];
      if (ts.isLabeledStatement(stmt)) {
        const labelName = stmt.label.text;
        const macro = import_core.globalRegistry.getLabeledBlock(labelName);
        if (macro) {
          if (this.verbose) {
            console.log(
              `[typemacro] Expanding labeled block macro: ${labelName}:`
            );
          }
          let continuation;
          if (macro.continuationLabels && i + 1 < statements.length) {
            const next = statements[i + 1];
            if (ts.isLabeledStatement(next) && macro.continuationLabels.includes(next.label.text)) {
              continuation = next;
              i++;
            }
          }
          try {
            const result = macro.expand(this.ctx, stmt, continuation);
            const expanded = Array.isArray(result) ? result : [result];
            for (const s of expanded) {
              const visited2 = ts.visitNode(s, this.visit.bind(this));
              if (visited2) {
                if (Array.isArray(visited2)) {
                  newStatements.push(
                    ...visited2.filter(ts.isStatement)
                  );
                } else {
                  newStatements.push(visited2);
                }
              }
            }
          } catch (error) {
            this.ctx.reportError(
              stmt,
              `Labeled block macro expansion failed: ${error}`
            );
            newStatements.push(
              this.createMacroErrorStatement(
                `typemacro: labeled block '${labelName}:' expansion failed: ${error}`
              )
            );
          }
          modified = true;
          continue;
        }
      }
      const visited = ts.visitNode(stmt, this.visit.bind(this));
      if (visited) {
        if (Array.isArray(visited)) {
          newStatements.push(...visited.filter(ts.isStatement));
        } else {
          newStatements.push(visited);
        }
      }
    }
    const factory = this.ctx.factory;
    if (ts.isSourceFile(node)) {
      return factory.updateSourceFile(node, newStatements);
    } else if (ts.isBlock(node)) {
      return factory.updateBlock(node, newStatements);
    } else {
      return factory.updateModuleBlock(node, newStatements);
    }
  }
  /**
   * Try to transform a node if it's a macro invocation
   */
  tryTransform(node) {
    if (ts.isCallExpression(node)) {
      const result = this.tryExpandExpressionMacro(node);
      if (result !== void 0) {
        return result;
      }
    }
    if (this.hasDecorators(node)) {
      const result = this.tryExpandAttributeMacros(node);
      if (result !== void 0) {
        return result;
      }
    }
    if (ts.isTaggedTemplateExpression(node)) {
      const result = this.tryExpandTaggedTemplate(node);
      if (result !== void 0) {
        return result;
      }
    }
    if (ts.isTypeReferenceNode(node)) {
      const result = this.tryExpandTypeMacro(node);
      if (result !== void 0) {
        return result;
      }
    }
    if (ts.isCallExpression(node) && ts.isPropertyAccessExpression(node.expression)) {
      const result = this.tryRewriteExtensionMethod(node);
      if (result !== void 0) {
        return result;
      }
    }
    return void 0;
  }
  hasDecorators(node) {
    return ts.canHaveDecorators(node) && ts.getDecorators(node) !== void 0;
  }
  tryExpandExpressionMacro(node) {
    let macroName;
    let identNode;
    if (ts.isIdentifier(node.expression)) {
      macroName = node.expression.text;
      identNode = node.expression;
    } else if (ts.isPropertyAccessExpression(node.expression)) {
      if (ts.isIdentifier(node.expression.expression)) {
        if (node.expression.expression.text === "macro") {
          macroName = node.expression.name.text;
          identNode = node.expression.name;
        } else {
          macroName = node.expression.name.text;
          identNode = node.expression;
        }
      }
    }
    if (!macroName || !identNode) return void 0;
    const macro = this.resolveMacroFromSymbol(
      identNode,
      macroName,
      "expression"
    );
    if (!macro) return void 0;
    if (this.verbose) {
      console.log(`[typemacro] Expanding expression macro: ${macroName}`);
    }
    try {
      const result = macro.expand(this.ctx, node, node.arguments);
      return ts.visitNode(result, this.visit.bind(this));
    } catch (error) {
      this.ctx.reportError(node, `Macro expansion failed: ${error}`);
      return this.createMacroErrorExpression(
        `typemacro: expansion of '${macroName}' failed: ${error}`
      );
    }
  }
  tryExpandAttributeMacros(node) {
    const decorators = ts.getDecorators(node);
    if (!decorators || decorators.length === 0) return void 0;
    let currentNode = node;
    const extraStatements = [];
    const remainingDecorators = [];
    let wasTransformed = false;
    for (const decorator of decorators) {
      const { macroName, args, identNode } = this.parseDecorator(decorator);
      if (macroName === "derive") {
        const deriveMacroResolved = identNode ? this.resolveMacroFromSymbol(identNode, "derive", "attribute") : import_core.globalRegistry.getAttribute("derive");
        if (deriveMacroResolved) {
          const derives = this.expandDeriveDecorator(decorator, node, args);
          if (derives) {
            extraStatements.push(...derives);
            wasTransformed = true;
            continue;
          }
        }
      }
      const macro = identNode ? this.resolveMacroFromSymbol(identNode, macroName, "attribute") : import_core.globalRegistry.getAttribute(macroName);
      if (macro) {
        if (this.verbose) {
          console.log(`[typemacro] Expanding attribute macro: ${macroName}`);
        }
        try {
          const result = macro.expand(
            this.ctx,
            decorator,
            currentNode,
            args
          );
          if (Array.isArray(result)) {
            if (result.length > 0) {
              currentNode = result[0];
              extraStatements.push(...result.slice(1).filter(ts.isStatement));
            }
          } else {
            currentNode = result;
          }
          wasTransformed = true;
        } catch (error) {
          this.ctx.reportError(
            decorator,
            `Attribute macro expansion failed: ${error}`
          );
          extraStatements.push(
            this.createMacroErrorStatement(
              `typemacro: attribute macro '${macroName}' failed: ${error}`
            )
          );
          remainingDecorators.push(decorator);
          wasTransformed = true;
        }
      } else {
        remainingDecorators.push(decorator);
      }
    }
    if (!wasTransformed) return void 0;
    if (remainingDecorators.length !== decorators.length) {
      currentNode = this.updateDecorators(currentNode, remainingDecorators);
    }
    const visited = ts.visitNode(currentNode, this.visit.bind(this));
    if (extraStatements.length > 0) {
      return [visited, ...extraStatements];
    }
    return visited;
  }
  parseDecorator(decorator) {
    const expr = decorator.expression;
    if (ts.isIdentifier(expr)) {
      return { macroName: expr.text, args: [], identNode: expr };
    }
    if (ts.isCallExpression(expr)) {
      if (ts.isIdentifier(expr.expression)) {
        return {
          macroName: expr.expression.text,
          args: Array.from(expr.arguments),
          identNode: expr.expression
        };
      }
    }
    return { macroName: "", args: [], identNode: void 0 };
  }
  expandDeriveDecorator(decorator, node, args) {
    if (!ts.isInterfaceDeclaration(node) && !ts.isClassDeclaration(node) && !ts.isTypeAliasDeclaration(node)) {
      this.ctx.reportError(
        decorator,
        "@derive can only be applied to interfaces, classes, or type aliases"
      );
      return void 0;
    }
    const statements = [];
    const typeInfo = this.extractTypeInfo(node);
    for (const arg of args) {
      if (!ts.isIdentifier(arg)) {
        this.ctx.reportError(arg, "derive arguments must be identifiers");
        continue;
      }
      const deriveName = arg.text;
      const macro = import_core.globalRegistry.getDerive(deriveName);
      if (!macro) {
        this.ctx.reportError(arg, `Unknown derive macro: ${deriveName}`);
        continue;
      }
      if (this.verbose) {
        console.log(`[typemacro] Expanding derive macro: ${deriveName}`);
      }
      try {
        const result = macro.expand(this.ctx, node, typeInfo);
        statements.push(...result);
      } catch (error) {
        this.ctx.reportError(arg, `Derive macro expansion failed: ${error}`);
      }
    }
    return statements.length > 0 ? statements : void 0;
  }
  extractTypeInfo(node) {
    const name = node.name?.text ?? "Anonymous";
    const type = this.ctx.typeChecker.getTypeAtLocation(node);
    const typeParameters = node.typeParameters ? Array.from(node.typeParameters) : [];
    const fields = [];
    const properties = this.ctx.typeChecker.getPropertiesOfType(type);
    for (const prop of properties) {
      const declarations = prop.getDeclarations();
      if (!declarations || declarations.length === 0) continue;
      const decl = declarations[0];
      const propType = this.ctx.typeChecker.getTypeOfSymbolAtLocation(
        prop,
        decl
      );
      const propTypeString = this.ctx.typeChecker.typeToString(propType);
      const optional = (prop.flags & ts.SymbolFlags.Optional) !== 0;
      const readonly = ts.isPropertyDeclaration(decl) || ts.isPropertySignature(decl) ? decl.modifiers?.some(
        (m) => m.kind === ts.SyntaxKind.ReadonlyKeyword
      ) ?? false : false;
      fields.push({
        name: prop.name,
        typeString: propTypeString,
        type: propType,
        optional,
        readonly
      });
    }
    return { name, fields, typeParameters, type };
  }
  tryExpandTaggedTemplate(node) {
    if (!ts.isIdentifier(node.tag)) return void 0;
    const tagName = node.tag.text;
    const taggedMacro = this.resolveMacroFromSymbol(
      node.tag,
      tagName,
      "tagged-template"
    );
    if (taggedMacro) {
      if (this.verbose) {
        console.log(`[typemacro] Expanding tagged template macro: ${tagName}`);
      }
      try {
        if (taggedMacro.validate && !taggedMacro.validate(this.ctx, node)) {
          this.ctx.reportError(
            node,
            `Tagged template validation failed for '${tagName}'`
          );
          return this.createMacroErrorExpression(
            `typemacro: tagged template '${tagName}' validation failed`
          );
        }
        const result = taggedMacro.expand(this.ctx, node);
        return ts.visitNode(result, this.visit.bind(this));
      } catch (error) {
        this.ctx.reportError(
          node,
          `Tagged template macro expansion failed: ${error}`
        );
        return this.createMacroErrorExpression(
          `typemacro: tagged template '${tagName}' expansion failed: ${error}`
        );
      }
    }
    const exprMacro = this.resolveMacroFromSymbol(
      node.tag,
      tagName,
      "expression"
    );
    if (!exprMacro) return void 0;
    if (this.verbose) {
      console.log(
        `[typemacro] Expanding tagged template via expression macro: ${tagName}`
      );
    }
    try {
      const result = exprMacro.expand(
        this.ctx,
        node,
        [node.template]
      );
      return ts.visitNode(result, this.visit.bind(this));
    } catch (error) {
      this.ctx.reportError(
        node,
        `Tagged template macro expansion failed: ${error}`
      );
      return this.createMacroErrorExpression(
        `typemacro: tagged template '${tagName}' expansion failed: ${error}`
      );
    }
  }
  tryExpandTypeMacro(node) {
    let macroName;
    let identNode;
    if (ts.isIdentifier(node.typeName)) {
      macroName = node.typeName.text;
      identNode = node.typeName;
    } else if (ts.isQualifiedName(node.typeName)) {
      if (ts.isIdentifier(node.typeName.left) && node.typeName.left.text === "typemacro") {
        macroName = node.typeName.right.text;
        identNode = node.typeName;
      }
    }
    if (!macroName || !identNode) return void 0;
    const macro = this.resolveMacroFromSymbol(identNode, macroName, "type");
    if (!macro) return void 0;
    if (this.verbose) {
      console.log(`[typemacro] Expanding type macro: ${macroName}`);
    }
    try {
      const typeArgs = node.typeArguments ? Array.from(node.typeArguments) : [];
      const result = macro.expand(this.ctx, node, typeArgs);
      return ts.visitNode(result, this.visit.bind(this));
    } catch (error) {
      this.ctx.reportError(node, `Type macro expansion failed: ${error}`);
      return node;
    }
  }
  /**
   * Try to rewrite an implicit extension method call.
   */
  tryRewriteExtensionMethod(node) {
    const propAccess = node.expression;
    const methodName = propAccess.name.text;
    const receiver = propAccess.expression;
    if (ts.isCallExpression(receiver) && ts.isIdentifier(receiver.expression)) {
      const calleeName = receiver.expression.text;
      const calleeMacro = this.resolveMacroFromSymbol(
        receiver.expression,
        calleeName,
        "expression"
      );
      if (calleeMacro) {
        return void 0;
      }
    }
    const receiverType = this.ctx.typeChecker.getTypeAtLocation(receiver);
    const existingProp = receiverType.getProperty(methodName);
    if (existingProp) {
      return void 0;
    }
    const typeName = this.ctx.typeChecker.typeToString(receiverType);
    let extension = import_core.globalExtensionRegistry.find(methodName, typeName);
    if (!extension) {
      const baseTypeName = typeName.replace(/<.*>$/, "");
      if (baseTypeName !== typeName) {
        extension = import_core.globalExtensionRegistry.find(methodName, baseTypeName);
      }
    }
    if (!extension) {
      return void 0;
    }
    if (this.verbose) {
      console.log(
        `[typemacro] Rewriting implicit extension: ${typeName}.${methodName}() \u2192 ${extension.typeclassName}.summon<${typeName}>("${typeName}").${methodName}(...)`
      );
    }
    const receiverText = receiver.getText ? receiver.getText() : ts.createPrinter().printNode(ts.EmitHint.Expression, receiver, this.ctx.sourceFile);
    const extraArgs = Array.from(node.arguments).map(
      (a) => a.getText ? a.getText() : ts.createPrinter().printNode(ts.EmitHint.Expression, a, this.ctx.sourceFile)
    ).join(", ");
    const allArgs = extraArgs ? `${receiverText}, ${extraArgs}` : receiverText;
    const code = `${extension.typeclassName}.summon<${typeName}>("${typeName}").${methodName}(${allArgs})`;
    try {
      const result = this.ctx.parseExpression(code);
      return ts.visitNode(result, this.visit.bind(this));
    } catch (error) {
      this.ctx.reportError(node, `Extension method rewrite failed: ${error}`);
      return void 0;
    }
  }
  createMacroErrorExpression(message) {
    const factory = this.ctx.factory;
    return factory.createCallExpression(
      factory.createParenthesizedExpression(
        factory.createArrowFunction(
          void 0,
          void 0,
          [],
          void 0,
          factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
          factory.createBlock([
            factory.createThrowStatement(
              factory.createNewExpression(
                factory.createIdentifier("Error"),
                void 0,
                [factory.createStringLiteral(message)]
              )
            )
          ])
        )
      ),
      void 0,
      []
    );
  }
  createMacroErrorStatement(message) {
    const factory = this.ctx.factory;
    return factory.createThrowStatement(
      factory.createNewExpression(
        factory.createIdentifier("Error"),
        void 0,
        [factory.createStringLiteral(message)]
      )
    );
  }
  updateDecorators(node, decorators) {
    const modifiers = decorators.length > 0 ? decorators : void 0;
    const factory = this.ctx.factory;
    if (ts.isClassDeclaration(node)) {
      return factory.updateClassDeclaration(
        node,
        modifiers ? [
          ...modifiers,
          ...node.modifiers?.filter((m) => !ts.isDecorator(m)) ?? []
        ] : node.modifiers?.filter((m) => !ts.isDecorator(m)),
        node.name,
        node.typeParameters,
        node.heritageClauses,
        node.members
      );
    }
    if (ts.isFunctionDeclaration(node)) {
      return factory.updateFunctionDeclaration(
        node,
        modifiers ? [
          ...modifiers,
          ...node.modifiers?.filter((m) => !ts.isDecorator(m)) ?? []
        ] : node.modifiers?.filter((m) => !ts.isDecorator(m)),
        node.asteriskToken,
        node.name,
        node.typeParameters,
        node.parameters,
        node.type,
        node.body
      );
    }
    if (ts.isMethodDeclaration(node)) {
      return factory.updateMethodDeclaration(
        node,
        modifiers ? [
          ...modifiers,
          ...node.modifiers?.filter((m) => !ts.isDecorator(m)) ?? []
        ] : node.modifiers?.filter((m) => !ts.isDecorator(m)),
        node.asteriskToken,
        node.name,
        node.questionToken,
        node.typeParameters,
        node.parameters,
        node.type,
        node.body
      );
    }
    if (ts.isInterfaceDeclaration(node)) {
      return node;
    }
    return node;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MacroTransformer
});
